### PR TITLE
feat: support for variable interpolation in workspace.* (in PipelineRun and TaskRun)

### LINKED ISSUE #7667: Support for variable interpolation in workspace.configmap.items * (in PipelineRun and TaskRun)
This is follow up issue of https://github.com/tektoncd/pipeline/issues/6384 

Currently, we do support workspace provided as

```
    - name: custom4
      configMap:
        name: $(params.name)-configmap
        items:
          - key: message
            path: my-message.txt 
```

but we do not support workspace provided as

```
    - name: custom4
      configMap:
        name: $(params.name)-configmap
        items:
          - key: message
            path: $(params.name)-message.txt 
 ```

This is to add the support for param replacement at items level in workspace

### PR DESCRIPTION
**Pull request by l-qing**
fix #7667

Support for variable interpolation in:

* spec.workspaces[].configMap.items[].*
* spec.workspaces[].secret.items[].*
* spec.workspaces[].projected.sources[].configMap.items[].*
* spec.workspaces[].projected.sources[].secret.items[].*

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
feat: support for variable interpolation in workspace.* (in PipelineRun and TaskRun)
```

/kind feature

### CODE REVIEW

File: docs/variables.md
Comment (l-qing) [inline docs/variables.md:88]: It should actually be camel case notation.

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (l-qing) [inline pkg/reconciler/pipelinerun/pipelinerun_test.go:4894]: Remove trailing whitespace.

File: pkg/reconciler/pipelinerun/resources/apply.go
Comment (l-qing) [inline pkg/reconciler/pipelinerun/resources/apply.go:400]: Refactor this section of the code so that it can be reused in taskrun and pipelinerun.

File: pkg/workspace/apply.go
Comment (l-qing) [inline pkg/workspace/apply.go:281]: This is an extracted public method for reuse in both taskrun and pipelinerun.

**Review by JeromeJu — COMMENTED** [review]
Thanks @l-qing for this! 
Would you mind splitting up the PR from refactoring and changing the variable camel cases according to https://github.com/tektoncd/community/blob/main/standards.md#small-pull-requests?

**Comment by l-qing** [comment]
/hold

Hold this PR until the #7701 is merged.

**Comment by l-qing** [comment]
> Thanks @l-qing for this! Would you mind splitting up the PR from refactoring and changing the variable camel cases according to https://github.com/tektoncd/community/blob/main/standards.md#small-pull-requests?

Thank you for your suggestion.
I've split the document modification PR #7701; please review it when you have time.

File: docs/variables.md
Comment (afrittoli) [reply docs/variables.md:88]: This will go away once this is rebased on top of the related PR, right?

Comment (l-qing) [reply docs/variables.md:88]: Yes, that's why I set up the `do-not-merge/hold`.

**Review by afrittoli — COMMENTED** [review]
Thanks! 
/approve

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (l-qing) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:6059]: I found that the latter part of the unit tests fixed the name of the PR, so any modifications would have to be changed consistently.

https://github.com/tektoncd/pipeline/blob/d27e6cf92dee58d364fe6183509d6f7810ec2971/pkg/reconciler/pipelinerun/pipelinerun_test.go#L6135-L6140

I've uniformly renamed `test-pipeline-run-different-service-accs` to `test-pipeline-run-variable-substitution` to better reflect the purpose of this case.

**Review by vdemeester — APPROVED** [review]
/lgtm

**Comment by l-qing** [comment]
> /lgtm

Thank you very much, I was just about to at you for comment. 😆

**Pull request merged**
